### PR TITLE
patch URLFilters to prevent `//`

### LINKED ISSUE #6057: Handle '//' (double slash) caused by baseurl: '/'
- [ ] I believe this to be a bug, not a question about using Jekyll.
- [x] I updated to the latest Jekyll (or) if on GitHub Pages to the latest `github-pages`
- [x] I read the CONTRIBUTION file at https://jekyllrb.com/docs/contributing/
- [x] This is a feature request.

---

- [x] I am on (or have tested on) ***macOS*** 10+

## My Reproduction Steps

To start off with, I realise that setting baseurl to `/` is the wrong way to do it, but in reality, it seems it is often done, and sometimes seems to be the only way to workaround other bugs/things:
* https://github.com/jekyll/jekyll-seo-tag/issues/200
* https://github.com/github/pages-gem/issues/359#issuecomment-263375327
* https://github.com/jekyll/github-metadata/issues/87
* etc

I figured that with the [relative/absolute filters added in 3.3](https://jekyllrb.com/news/#2-relative_url-and-absolute_url-filters) there was now a standard/universal place to fix this in a way that would make the end user experience easier, while possibly also hinting them towards not using the 'bad method'

```
# _config.yml
baseurl: "/"

# default.html
<h1>{{ "/docs/assets/" | relative_url }}</h1>
<h1>{{ "/docs/assets/" | absolute_url }}</h1>
```

Outputs
```
<h1>//docs/assets/</h1>
<h1>http://localhost:4000//docs/assets/</h1>
```

## The Output I Wanted

```
<h1>/docs/assets/</h1>
<h1>http://localhost:4000/docs/assets/</h1>
```

This could be achieved by using a regex/similar to strip off any trailing/leading slashes before concatenation.

For what it's worth, the hacky filter I used in my theme to work around this was:

```
module Jekyll
  module ChompSlashFilter
    def chomp_slash(input)
      input.chomp('/')
    end
  end
end

Liquid::Template.register_filter(Jekyll::ChompSlashFilter)
```

But this obviously doesn't let me work around it when it comes up in other plugins.

It would also be reasonable to expect a warning from Jekyll about how using `/` in baseurl is a bad idea/why, and how to do it better/properly (maybe a link to a blog post about it or similar).

### PR DESCRIPTION
**Pull request by ashmaroli**
Fix #6057 

/cc @jekyll/ecosystem 
requesting additional review from @parkr as well.

### CODE REVIEW

File: lib/jekyll/filters/url_filters.rb [hunk 1]
```diff
@@ -23,14 +22,23 @@ def absolute_url(input)
...
 
       private
+
+      def site
+        @context.registers[:site]
+      end
+
+      def sanitized_baseurl
+        baseurl = site.config["baseurl"]
+        return baseurl unless baseurl.nil? || (baseurl == "/")
```
Comment (alias1) [inline lib/jekyll/filters/url_filters.rb]: Will this fail and not return if baseurl is nil? (Unsure of ruby specifics for unless)

What happens if baseurl is set to /foo/bar/ or even something really silly like "///////" or /foo/bar//" etc?

Can we compress trailing slashes to a single one, or removed entirely, as required by the concatenation?

File: test/test_filters.rb [hunk 2]
```diff
@@ -448,6 +457,15 @@ def select; end
         })
         assert_equal "/base", filter.relative_url(page_url)
       end
+
+      should "not prepend a forward slash if both input and baseurl are simply '/'" do
```
Comment (alias1) [inline test/test_filters.rb]: Should this be "append"?

File: lib/jekyll/filters/url_filters.rb (hunk 1 above)
Comment (alias1) [inline lib/jekyll/filters/url_filters.rb]: And I guess I haven't really checked what currently happens, but what if similarly there is some silliness going on in the site URL part with slashes. Can we robustly handle this in both potential locations?

Comment (pathawks) [inline lib/jekyll/filters/url_filters.rb]: > What happens if baseurl is set to /foo/bar/ or even something really silly like "///////" or /foo/bar//" etc?

Pray, Mr. Babbage, if you put into the machine wrong figures, will the right answers come out?

**Comment by alias1** [comment]
A fair point, but I suppose my use cases there were exaggerated examples to highlight a point. Should we be catering to a specific instance of a problem (e.g. Baseurl = /), or the more general pattern of the problem (ending in 1 or more slashes, that when concatenated cause duplication)

My personal view leans towards fixing the pattern rather than the specific instance.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by pathawks** [comment]
But multiple slashes are perfectly legal in a URL. Do we really want to stop supporting that just so setting  `baseurl: "////////"` has no effect?

Let's try to figure out what the root problem is and try to fix that with as few unintended consequences as possible, rather than trying to decide how hypothetical situations should be magically changed without the users permission.

File: lib/jekyll/filters/url_filters.rb (hunk 1 above)
Comment (ashmaroli) [inline lib/jekyll/filters/url_filters.rb]: > Will this fail and not return if baseurl is nil? 

Ruby will `return nil` if `baseurl` equals `nil` or `/`

> Can we compress trailing slashes to a single one, or removed entirely, as required by the concatenation?

IMO substituting a regex match with something is not necessary as it is known to affect performance.  But again, there is no issue in employing that method if there's a strong use case. Setting `baseurl` with multiple slashes is wrong.
The least appropriate measure would be to output a warning message that the `baseurl` has been configured incorrectly.

**Comment by ashmaroli** [comment]
updated to return valid url if `baseurl == /base/`

**Comment by parkr** [comment]
@jekyllbot: merge +bug

**Comment by ashmaroli** [comment]
Would it make sense to partially revert this (remove the private method `:site` introduced here) instead of fixing plugins downstream?

/cc @parkr @benbalter @pathawks

**Comment by parkr** [comment]
> Would it make sense to partially revert this (remove the private method :site introduced here) instead of fixing plugins downstream?

Making this more backwards-compatible sounds like a great idea!

**Pull request merged**